pkg/controller/status/status.go

```diff
@@ -639,3 +669,35 @@ func (m *statusManager) degradedReason() string {
 	}
 	return strings.Join(reasons, "; ")
 }
+
+func hasPendingCSR(ctx context.Context, cli client.Client, label string) (bool, error) {
+	csrs := &v1beta1.CertificateSigningRequestList{}
+	lbls := map[string]string{
+		"k8s-app": label,
+	}
+	selector := labels.SelectorFromSet(lbls)
+	err := cli.List(ctx, csrs, &client.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return false, err
+	}
+
+	if len(csrs.Items) == 0 {
+		return false, nil
+	}
+
+	for _, csr := range csrs.Items {
+		if len(csr.Status.Conditions) == 0 {
+			// No conditions means status is pending
+			return true, nil
+		}
+		// no condition approved, means it is pending.
+		for _, condition := range csr.Status.Conditions {
+			if condition.Status == v1.ConditionUnknown {
+				return true, nil
+			} else if condition.Type == v1beta1.CertificateApproved && csr.Status.Certificate == nil {
```

Comment:
Not sure I full understand this - what is the series of states that a CSR goes through? Could we simply say that if `csr.Status.Certificate` is nil that the signing procedure hasn't completed yet?

Reply:
Excellent question. I think my docs change is the best answer to your question as I made visible what the pod statuses look like: https://docs.tigera.io/master/security/comms/certificate-management

TLDR: two separate endpoints are necessary to go from pending to completely done:
- Add certificate
- Approve
While the docs state that it should be first approved before signing, in practice this can happen in any order.

Reply:
Lastly, to give a better answer to your question It is possible that a CSR is rejected/denied or if there is another non-approved condition. In that case there is not a certificate and at the same time, the CSR is not pending (nor should the tigerastatus). The init container will then crash, delete the CSR and try again on next init until it is approved and signed.